lib/spack/spack/solver/asp.py

```diff
@@ -763,6 +759,30 @@ def key_fn(version):
                 )
             )
 
+        for v in most_to_least_preferred:
+            # There are two paths for creating the ref_version in GitVersions.
+            # The first uses a lookup to supply a tag and distance as a version.
+            # The second is user specified and can be resolved as a standard version.
+            # This second option is constrained such that the user version must be known to Spack
+            # for now we distinguish this by if the ref_version is a string
+            if isinstance(v.version, spack.version.GitVersion) and isinstance(
+                v.version.ref_version, str
+            ):
+                ref_version = spack.version.Version(v.version.ref_version)
+                self.gen.fact(fn.version_equivalent(pkg.name, v.version, ref_version))
+                # disqualify any git supplied version from user if they weren't already known
+                # versions in spack
+                if not any(ref_version == dv.version for dv in most_to_least_preferred):
+                    msg = (
+                        "The reference version '{version}' for package '{package}' is not defined."
+                        " Either choose another reference version or define '{version}' in your"
+                        " version preferences or package.py file for {package}.".format(
+                            package=pkg.name, version=v.version.ref_version
+                        )
+                    )
+
+                    raise UnsatisfiableSpecError(msg)
```

Comment:
I think this has a problem -- this is applying to all git versions, not just the ones that have been directly associated

Reply:
Not any more.  That is what the `isinstance(v.version.ref_version, str)` in L768-769 is taking care of.  The type is tuple for `GitVersions` without the manually paired reference version.